entoas/view.go

```diff
@@ -0,0 +1,247 @@
+package entoas
+
+import (
+	"fmt"
+	"strings"
+
+	"entgo.io/contrib/entoas/serialization"
+	"entgo.io/ent/entc/gen"
+	"github.com/stoewer/go-strcase"
+)
+
+type (
+	// A View is a subset of a gen.Type. It may hold fewer Fields and Edges than the gen.Type it is derived from.
+	View struct {
+		Type   *gen.Type
+		Fields []*ViewField
+		Edges  []*ViewEdge
+	}
+	ViewField struct {
+		*gen.Field
+	}
+	ViewEdge struct {
+		*gen.Edge
+	}
+)
+
+// Views returns all views that are needed to fill the OAS.
+func Views(g *gen.Graph) (map[string]*View, error) {
+	m := make(map[string]*View)
+	for _, n := range g.Nodes {
+		ops, err := NodeOperations(n)
+		if err != nil {
+			return nil, err
+		}
+		// For every operation add a schema to use.
+		for _, op := range ops {
+			// Skip the delete operation (of course).
+			if op == OpDelete {
+				continue
+			}
+			gs, err := groupsForOperation(n.Annotations, op)
+			if err != nil {
+				return nil, err
+			}
+			v, err := view(n, gs)
+			if err != nil {
+				return nil, err
+			}
+			vn, err := viewName(n, op)
+			if err != nil {
+				return nil, err
+			}
+			m[vn] = v
+			// Collect the "tree" of edges to load on this node and operation.
+			es, err := EdgeTree(n, gs)
+			if err != nil {
+				return nil, err
+			}
+			// Flatten the tree. Create a view for every type involved.
+			for _, e := range es.Flatten() {
+				v, err := view(e.Type, gs)
+				if err != nil {
+					return nil, err
+				}
+				evn, err := viewNameEdge(vn, e)
+				if err != nil {
+					return nil, err
+				}
+				m[evn] = v
+			}
+		}
+		// Look at the edges and the operations exposed on them. Do create the views too.
+		for _, e := range n.Edges {
+			ops, err := EdgeOperations(e)
+			if err != nil {
+				return nil, err
+			}
+			// For every operation add a schema to use.
+			for _, op := range ops {
+				// Skip the delete operation (of course).
+				if op == OpDelete {
+					continue
+				}
+				gs, err := groupsForOperation(e.Annotations, op)
+				if err != nil {
+					return nil, err
+				}
+				v, err := view(e.Type, gs)
+				if err != nil {
+					return nil, err
+				}
+				vn, err := edgeViewName(n, e, op)
+				if err != nil {
+					return nil, err
+				}
+				m[vn] = v
+				// Collect the "tree" of edges to load on this edge and operation.
+				es, err := EdgeTree(n, gs)
+				if err != nil {
+					return nil, err
+				}
+				// Flatten the tree. Create a view for every type involved.
+				for _, t := range es.Flatten() {
+					v, err := view(t.Type, gs)
+					if err != nil {
+						return nil, err
+					}
+					evn, err := viewNameEdge(vn, t)
+					if err != nil {
+						return nil, err
+					}
+					m[evn] = v
+				}
```

Comment:
Consider moving the duplicate blocks to a function.

Reply:
Makes it more complex to gain very little and lose readability imho. There are just minor differences between those too. I think it should stay this way.